### PR TITLE
Implement prediction logging in standard transformer

### PR DESCRIPTION
**Pull request by tiopramayudi**
**What this PR does / why we need it**:

As part of closed loop ml initiative, merlin will publish prediction log through standard transformer. Some of modifications related to this PR
* Adding kafka publisher implementation in standard transformer
* Adding new prediction log operation in standard transformer, this enables us to also do validation before the deployment
* UI implementation to enable prediction log configuration
* API & SDK implementation to enable prediction log configuration
* Generated code for go and python (for sdk purpose)
**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note

```

https://user-images.githubusercontent.com/2369255/228119751-c3d87a88-117c-44da-8590-1a4a7c43d2f0.mov

**Checklist**

- [x] Added unit test, integration, and/or e2e tests
- [x] Tested locally
- [ ] Updated documentation
- [x] Update Swagger spec if the PR introduce API changes
- [x] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

File: api/client/examples/examples/model-endpoints/main.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	"context"
+	"log"
+	"net/http"
+	"os"
+
+	"github.com/gojek/merlin/client"
+	"golang.org/x/oauth2/google"
+)
+
+func main() {
```
Comment (pradithya) [inline api/client/examples/examples/model-endpoints/main.go:13]: What's the purpose of this source code?

File: api/client/examples/examples/project-endpoints/main.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	"context"
+	"log"
+	"net/http"
+	"os"
+
+	"github.com/gojek/merlin/client"
+	"golang.org/x/oauth2/google"
+)
+
+func main() {
```
Comment (pradithya) [inline api/client/examples/examples/project-endpoints/main.go:13]: Same question? Is this file necessary?

File: api/config/config.go [hunk 3]
```diff
@@ -280,6 +281,16 @@ type StandardTransformerConfig struct {
 	DefaultFeastSource spec.ServingSource `envconfig:"DEFAULT_FEAST_SOURCE" default:"BIGTABLE"`
 	Jaeger             JaegerConfig
 	SimulationFeast    SimulationFeastConfig
+	Kafka              KafkaConfig
+}
+
+// Kafka configuration
```
Comment (pradithya) [inline api/config/config.go:287]: nit: Make the comment more specific `Kafka configuration for publishing prediction log`

File: api/models/logger.go [hunk 4]
```diff
@@ -45,12 +46,31 @@ func (l LoggerMode) isValid() bool {
...
+	DestinationURL string                  `json:"-"`
+	Model          *LoggerConfig           `json:"model"`
+	Transformer    *LoggerConfig           `json:"transformer"`
+	Prediction     *PredictionLoggerConfig `json:"prediction"`
 }
 
+// PredictionLoggerConfig holds information about prediction logger
+type PredictionLoggerConfig struct {
+	Enabled          bool   `json:"enabled"`
+	RawFeaturesTable string `json:"raw_features_table"`
```
Comment (pradithya) [inline api/models/logger.go:60]: `RawFeaturesTable` and `EntitiesTable` are currently not used right?

File: api/pkg/kafka/config.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+// Serialization format
+type SerializationFormat string
+
+const (
+	Protobuf SerializationFormat = "protobuf"
+	JSON     SerializationFormat = "json"
+)
+
+// Kafka configuration
+type Config struct {
```
Comment (pradithya) [inline api/pkg/kafka/config.go:12]: Why do we have `kafka.Config` and `config.KafkaConfig`. Can we remove the duplicate?

File: api/pkg/kafka/kafka.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+	upiv1 "github.com/caraml-dev/universal-prediction-interface/gen/go/grpc/caraml/upi/v1"
+	kafkalib "github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	kafkaConnectTimeoutMs = 1000
```
Comment (pradithya) [inline api/pkg/kafka/kafka.go:16]: I think it should be part of the kafka configuration

File: api/pkg/kafka/kafka.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+		}
+		msgValue = val
+	}
+	err := kp.producer.Produce(&kafkalib.Message{
+		TopicPartition: kafkalib.TopicPartition{
+			Topic:     &kp.topic,
+			Partition: kafkalib.PartitionAny,
+		},
+		Value: msgValue,
+		Key:   nil,
```
Comment (pradithya) [inline api/pkg/kafka/kafka.go:94]: Why don't we set the key?

File: api/pkg/transformer/pipeline/compiler.go [hunk 8]
```diff
@@ -37,33 +36,28 @@ type Compiler struct {
...
 		feastOptions:            feastOptions,
-		logger:                  logger,
-		operationTracingEnabled: tracingEnabled,
-		transformerValidationFn: validationFn,
-		jsonpathSourceType:      jsonpathSourceType,
+		operationTracingEnabled: false,
 	}
+
+	for _, opt := range opts {
+		opt(compiler)
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/compiler.go:57]: Neat!

File: api/pkg/transformer/pipeline/environment.go [hunk 9]
```diff
@@ -44,6 +44,10 @@ func (e *Environment) Preprocess(ctx context.Context, rawRequest types.Payload,
 	e.symbolRegistry.SetRawRequestHeaders(rawRequestHeaders)
 	e.SetOutput(rawRequest)
 
+	if !e.IsPreprocessOpExist() {
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/environment.go:47]: What's the purpose of this condition here? Won't `e.compiledPipeline.Preprocess(ctx, e)` return immediately if there is no preprocess operation?

File: api/pkg/transformer/pipeline/prediction_log_op.go [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+func (pl *PredictionLogOp) buildPredictionLog(predictionResult *types.PredictionResult, env *Environment) (*upiv1.PredictionLog, error) {
+
+	originalRequest := env.SymbolRegistry().RawRequest()
+	request, validReq := originalRequest.(*types.UPIPredictionRequest)
+	if !validReq {
+		return nil, fmt.Errorf("type of prediction request is not valid: %T", originalRequest)
+	}
+
+	upiRequest := &types.UPIPredictionRequest{}
+	if err := copier.CopyWithOption(upiRequest, request, copier.Option{IgnoreEmpty: true, DeepCopy: true}); err != nil {
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/prediction_log_op.go:53]: Why do we copy?

File: api/pkg/transformer/pipeline/upi_postprocess_output_op.go [hunk 11]
```diff
@@ -34,7 +35,10 @@ func (up *UPIPostprocessOutputOp) Execute(ctx context.Context, env *Environment)
 		return fmt.Errorf("not valid type %T", modelResponse)
 	}
 
-	copiedResponse := upiModelResponse
+	copiedResponse := &types.UPIPredictionResponse{}
+	if err := copier.CopyWithOption(copiedResponse, upiModelResponse, copier.Option{IgnoreEmpty: true, DeepCopy: true}); err != nil {
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/upi_postprocess_output_op.go:39]: Why do we copy in here?

File: api/pkg/transformer/pipeline/upi_preprocess_output_op.go [hunk 12]
```diff
@@ -51,7 +52,10 @@ func (up *UPIPreprocessOutputOp) Execute(ctx context.Context, env *Environment)
 
 	transformerInput := &upiv1.TransformerInput{}
 	transformerInput.Tables = transformerInputTables
-	copiedRequest := enrichedRequest
+	copiedRequest := &types.UPIPredictionRequest{}
+	if err := copier.CopyWithOption(copiedRequest, enrichedRequest, copier.Option{IgnoreEmpty: true, DeepCopy: true}); err != nil {
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/upi_preprocess_output_op.go:56]: Why do we copy? We should minimize amount of copy to minimze the overhead introduced by standard transformer.

File: api/pkg/transformer/types/prediction_result.go [hunk 13]
```diff
@@ -0,0 +1,13 @@
+package types
+
+type PredictionResult struct {
```
Comment (pradithya) [inline api/pkg/transformer/types/prediction_result.go:3]: nit: can you add some comments to these new structs?

File: protos/merlin/transformer/spec/prediction_log.proto [hunk 14]
```diff
@@ -0,0 +1,11 @@
+syntax = "proto3";
+
+package merlin.transformer;
+
+option go_package = "github.com/gojek/merlin/pkg/transformer/spec";
+
+message PredictionLogConfig {
```
Comment (pradithya) [inline protos/merlin/transformer/spec/prediction_log.proto:7]: nit: Please add comments to the new proto message

**Review by pradithya — COMMENTED** [review]
Thanks! It's quite a comprehensive work. Can you add screenshots or recording of the UI to the PR description?

File: api/pkg/kafka/config.go (hunk 5 above)
Comment (tiopramayudi) [reply api/pkg/kafka/config.go:12]: initially only `kafka.Config` that existed, but due to that package relies on confluent kafka go, librdkafka must be installed also in merlin side which is unnecessary. Hence I created the new one `KafkaConfig`

File: api/client/examples/examples/model-endpoints/main.go (hunk 1 above)
Comment (tiopramayudi) [reply api/client/examples/examples/model-endpoints/main.go:13]: ah this one is generated, maybe I need to only pick the necessary files

File: api/pkg/transformer/pipeline/environment.go (hunk 9 above)
Comment (tiopramayudi) [reply api/pkg/transformer/pipeline/environment.go:47]: previously this one is called in `handler.go` to return immediately without doing `Unmarshaling` the request/response if there is no preprocess/postprocess . I move it to `environment.go` to handle case where user doesn't specify preprocess or postprocess ops, so when publish the prediction log the rawRequest can be fetched from `symbolRegistry` but I just realized by moving it to `environment.go` it defeats the purpose of that condition (to reduce the unmarshal when there is no preprocess or postprocess operation)

File: api/pkg/transformer/pipeline/upi_postprocess_output_op.go (hunk 11 above)
Comment (tiopramayudi) [reply api/pkg/transformer/pipeline/upi_postprocess_output_op.go:39]: yeah, i think this one unnecessary since if we do modification in predictor response is still ok to be mutated since it won't be referred in another operation. Will remove the copy

**Comment by ariefrahmansyah** [comment]
@tiopramayudi Tio, please fix merge conflicts 🙏🏻

**Comment by pradithya** [comment]
LGTM, thanks @tiopramayudi

File: api/service/version_endpoint_service.go [hunk 15]
```diff
@@ -172,6 +172,32 @@ func (k *endpointService) override(left *models.VersionEndpoint, right *models.V
...
 	}
 
+	// override logger
+	if right.Logger != nil {
+		left.Logger = right.Logger
+		left.Logger.DestinationURL = k.loggerDestinationURL
+		predictionLogger := left.Logger.Prediction
+		if predictionLogger != nil && right.Protocol == protocol.UpiV1 {
+			left.Logger.Prediction = predictionLogger
+		}
```
Comment (ariefrahmansyah) [inline api/service/version_endpoint_service.go:182]: I don't see a value in having these codes. There's no change in predictionLogger object

**Review by ariefrahmansyah — APPROVED** [review]
Other than that one nit comment, LGTM. Thanks @tiopramayudi !

File: api/service/version_endpoint_service.go (hunk 15 above)
Comment (tiopramayudi) [reply api/service/version_endpoint_service.go]: nice catch! Updated
